### DISCUSSION TITLE
upstream SSL certificate verify error: (2:unable to get issuer certificate) while SSL handshaking to upstream.

### DISCUSSION THREAD

**Discussion by GK722185** [nginx/nginx#409] [Q&A] [answered]
I have a reverse proxy server where the root certificate is configured using the proxy_ssl_trusted_certificate directive in Nginx.

The upstream server is configured with a certificate that includes the root certificate appended to it. I verified that there is no intermediate certificate involved in this setup.

However, all requests to the reverse proxy fail with a 502 error, and the logs show the following message:
upstream SSL certificate verify error: (2:unable to get issuer certificate) while SSL handshaking to upstream.

Additional Information:

The Subject DN and Issuer DN in the root certificate are different, but the public keys for the issuer and subject are identical.
If I add the root certificate to the trust store, browsers trust the server without issues. Similarly, adding the certificate to a Java keystore allows the server to be trusted. However, Nginx still does not trust the server and throws the same error.
Could the discrepancy between the Subject DN and Issuer DN in the root certificate be causing this issue? If not, what could be the potential root cause and how can I resolve it?

**Comment by pluknet** [comment, marked as answer]
> I have a reverse proxy server where the root certificate is configured using the proxy_ssl_trusted_certificate directive in Nginx.
> 
> The upstream server is configured with a certificate that includes the root certificate appended to it. I verified that there is no intermediate certificate involved in this setup.
> 
> However, all requests to the reverse proxy fail with a 502 error, and the logs show the following message: upstream SSL certificate verify error: (2:unable to get issuer certificate) while SSL handshaking to upstream.
> 
> Additional Information:
> 
> The Subject DN and Issuer DN in the root certificate are different

Which essentially means it is not self-signed, and OpenSSL cannot build a full chain for verification.

>  If I add the root certificate to the trust store, browsers trust the server without issues.

You can achieve the same result by explicitly marking your root certificate as "trusted", see openssl-req parameters `-addtrust`
and `-trustout` to add certificate trust auxiliary information.

See for additional details:
https://mailman.nginx.org/pipermail/nginx-devel/2022-May/6YBO7TZ27J7L75LAKADIV24WFFMSNLN5.html

**Reply by GK722185** [reply]
@pluknet Thanks for the info. But i have few queries regarding the Self Signed certificate.
> Which essentially means it is not self-signed, and OpenSSL cannot build a full chain for verification.


RFC 5280, in Section 4.1.2.4 (which outlines the "Subject" and "Issuer" fields of an X.509 certificate), does not explicitly say that the Issuer DN and Subject DN must be the same, but it strongly implies that self-signed certificates should have identical Issuer and Subject DNs because the certificate is issued by the same entity.
We can also verify the root certificate verifying the issuer and subject public key as the root itself signed it.

In addition to that, ssl_verify_partial_chain is used for client certificate authentication but here i'm facing issue with upstream server certificate verification. My reverse proxy nginx can't trust the upstream server certificate even i store the upstream's server rootCA certificate in nginx trust store using  proxy_ssl_trusted_certificate.

The main query i have here is, browser and java HttpsURLConnection trust the server if i add the root certificate in its trust store, but nginx can't. Kindly help me to understand how nginx fetch the root certificate and the steps to resolve this case.

**Reply by pluknet** [reply]
>RFC 5280, in Section 4.1.2.4 <..> does not explicitly say that the Issuer DN and Subject DN must be the same

RFC 5280, Section 3.2:
```Self-signed certificates are self-issued certificates```
This is followed by OpenSSL, see in particular x509v3_cache_extensions() and documentation on EXFLAG_SS.

The provided link is for certificate verification details in OpenSSL (start from "After reading...").
It may explain that Java and your browser do not supposedly build a full chain.
